core/build.go

```diff
@@ -46,14 +94,275 @@ func (b *build) hasFinished() bool {
+	b.m.Unlock()
+
+	// gets child processes killed, probably linux only
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	b.stdpipes = newStdpipes(stdout, stderr)
+	err = cmd.Start()
+	if err != nil {
+		cmd.Process.Kill()
+		return err
+	}
+	b.loginfof("successfully started build")
+	b.state = buildStateStarted
+
+runSyncLoop:
+	for {
+		select {
+		case <-b.stdpipes.Done:
+			err = cmd.Wait() // stdout/err have finished, just need to wait for the process to exit
+			if err != nil {
+				b.logwarnf("Build exited with non zero error code")
+				b.m.Lock()
+				defer b.m.Unlock()
+				b.exitCode = 1
+				b.cmd = nil
+				return err
+			}
+			break runSyncLoop
+
+		case <-time.After(config.Deadline):
+			b.logwarnf("Cancelling build as deadline reached")
+			err := b.Stop()
+			if err != nil {
+				b.logcritf("Couldn't stop build: %s", err)
+				b.m.Lock()
+				defer b.m.Unlock()
+				b.exitCode = 500
+				b.cmd = nil
+				return err
+			}
+		}
+	}
+
+	b.m.Lock()
+	defer b.m.Unlock()
+	b.cmd = nil
+
+	b.loginfof("Build finished")
+	return nil
+
 }
 
 // Start will start the given build, it will error with ErrAlreadyStarted if the build is already running
-func (b *build) Start() error { return nil }
+func (b *build) Start() error {
+	if b == nil {
+		return errors.New("b is nil")
+	}
+
+	b.m.RLock()
+
+	if b.state.HasStarted() || b.state.HasStopped() {
+		b.m.RUnlock()
+		return ErrProcessAlreadyStarted
+	}
+
+	b.state = buildStateWaitingForProvisioning
+
+	var config BuildConfig
+	config = *b.config
+	if config.Deadline < time.Duration(time.Millisecond) {
+		b.logwarnf("deadline not set in config, defaulting to 30 minutes")
+		config.Deadline = time.Minute * 30
+	}
+
+	b.m.RUnlock()
+
+	b.parentApp.SendEvent(fmt.Sprintf("/build/app:%s/started/token:%s", b.parentApp.Name(), b.Token()))
+	go func() {
+		err := b.runBuildSync(config)
+		if err != nil {
+			b.logwarnf("Build exited with error: %s", err)
+		}
+
+		// move artifacts over to perminent storage
+		// TODO get from config
+		perminentStorageDir := "/tmp/ngbuildartifacts/"
+		artifactDir := filepath.Join(perminentStorageDir, b.Token())
+		if err := os.MkdirAll(artifactDir, 0766); err != nil {
+			b.logcritf("Couldn't create artifact directory %s: %s", artifactDir, err)
+		} else {
+			// TODO, all this stuff, needs things defined in config
+			/*
+				for _, artifact := range b.parentApp.GetArtifactList() {
+					// copy file to artifacts directory
+				}
+			*/
+		}
+
+		b.parentApp.SendEvent(fmt.Sprintf("/build/app:%s/complete/token:%s", b.parentApp.Name(), b.Token()))
+	}()
+
+	return nil
+}
 
 // Stop will stop the given build, it will error with ErrAlreadyStopped if the build has finished
-func (b *build) Stop() error { return nil }
+func (b *build) Stop() error {
+	if b == nil {
+		return errors.New("b is nil")
+	}
+
+	b.m.RLock()
```

Comment:
Feels like just doing a defer b.m.Unlock here would make the code below a bit cleaner